File: Backend/src/main/java/org/example/edenmind/service/NotificationService.java
package org.example.edenmind.service;

import org.example.edenmind.entities.Notification;
import org.example.edenmind.entities.User;
import org.example.edenmind.repositories.NotificationRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

@Service
public class NotificationService {

    @Autowired
    private NotificationRepository notificationRepository;

    /**
     * Create a notification for a user
     */
    public Notification createNotification(User user, String title, String message, String type) {
        Notification notification = new Notification(user, title, message, type);
        return notificationRepository.save(notification);
    }

    /**
     * Notification when user logs their mood
     */
    public void notifyMoodLogged(User user, String mood) {
        String title = "Mood Logged! 📝";
        String message = "Great job logging your mood as '" + mood + "'. Keep tracking to see patterns in your wellness journey.";
        createNotification(user, title, message, "achievement");
    }

    /**
     * Notification for daily reminder
     */
    public void sendDailyReminder(User user) {
        String title = "Daily Check-in 🌟";
        String message = "Don't forget to log your mood today! Taking a moment to reflect can help improve your mental wellness.";
        createNotification(user, title, message, "reminder");
    }

    /**
     * Notification when user completes a meditation
     */
    public void notifyMeditationCompleted(User user, String sessionName, int minutes) {
        String title = "Meditation Complete! 🧘";
        String message = "You completed a " + minutes + "-minute session of '" + sessionName + "'. Great work on prioritizing your mental health!";
        createNotification(user, title, message, "achievement");
    }

    /**
     * Notification when user starts their first chat
     */
    public void notifyFirstChat(User user) {
        String title = "Welcome to ZenBot! 💬";
        String message = "Great job starting your first conversation with ZenBot. Feel free to share your thoughts anytime.";
        createNotification(user, title, message, "tip");
    }

    /**
     * Notification for wellness tip
     */
    public void sendWellnessTip(User user, String tip) {
        String title = "Wellness Tip 💡";
        createNotification(user, title, tip, "tip");
    }

    /**
     * Notification when user plays a therapeutic game
     */
    public void notifyGamePlayed(User user, String gameName) {
        String title = "Game Completed! 🎮";
        String message = "You finished playing '" + gameName + "'. These exercises can help reduce stress and improve mindfulness.";
        createNotification(user, title, message, "achievement");
    }

    /**
     * Notification for streak achievement
     */
    public void notifyStreak(User user, int days) {
        String title = "🔥 " + days + "-Day Streak!";
        String message = "Amazing! You've been using EdenMind for " + days + " days in a row. Keep up the great work!";
        createNotification(user, title, message, "achievement");
    }

    /**
     * Notification for new feature update
     */
    public void notifyUpdate(User user, String feature) {
        String title = "New Feature! 🆕";
        String message = "Check out the new " + feature + " feature in EdenMind!";
        createNotification(user, title, message, "update");
    }
}

File: app-v2/eden_mind_app/lib/features/notifications/notification_service.dart
import 'dart:convert';
import 'package:eden_mind_app/config/app_config.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:http/http.dart' as http;

class NotificationService {
  final FlutterSecureStorage _secureStorage;
  final http.Client _client;

  NotificationService({
    http.Client? client,
    FlutterSecureStorage? secureStorage,
  }) : _client = client ?? http.Client(),
       _secureStorage = secureStorage ?? const FlutterSecureStorage();

  String get _baseUrl => '${AppConfig.baseUrl}/notifications';

  Future<String?> _getToken() async {
    return await _secureStorage.read(key: 'jwt_token');
  }

  /// Get all notifications for the current user
  Future<List<Map<String, dynamic>>> getNotifications() async {
    final token = await _getToken();
    if (token == null) throw Exception('No authentication token');

    final response = await _client.get(
      Uri.parse(_baseUrl),
      headers: {'Authorization': 'Bearer $token'},
    );

    if (response.statusCode == 200) {
      final List<dynamic> data = jsonDecode(response.body);
      return data.cast<Map<String, dynamic>>();
    } else {
      throw Exception('Failed to load notifications');
    }
  }

  /// Get unread notification count
  Future<int> getUnreadCount() async {
    final token = await _getToken();
    if (token == null) return 0;

    try {
      final response = await _client.get(
        Uri.parse('$_baseUrl/unread-count'),
        headers: {'Authorization': 'Bearer $token'},
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return data['count'] ?? 0;
      }
    } catch (e) {
      // Silently fail for unread count
    }
    return 0;
  }

  /// Mark a notification as read
  Future<void> markAsRead(int notificationId) async {
    final token = await _getToken();
    if (token == null) throw Exception('No authentication token');

    final response = await _client.put(
      Uri.parse('$_baseUrl/$notificationId/read'),
      headers: {'Authorization': 'Bearer $token'},
    );

    if (response.statusCode != 200) {
      throw Exception('Failed to mark notification as read');
    }
  }

  /// Mark all notifications as read
  Future<void> markAllAsRead() async {
    final token = await _getToken();
    if (token == null) throw Exception('No authentication token');

    final response = await _client.put(
      Uri.parse('$_baseUrl/read-all'),
      headers: {'Authorization': 'Bearer $token'},
    );

    if (response.statusCode != 200) {
      throw Exception('Failed to mark all as read');
    }
  }

  /// Delete a notification
  Future<void> deleteNotification(int notificationId) async {
    final token = await _getToken();
    if (token == null) throw Exception('No authentication token');

    final response = await _client.delete(
      Uri.parse('$_baseUrl/$notificationId'),
      headers: {'Authorization': 'Bearer $token'},
    );

    if (response.statusCode != 204) {
      throw Exception('Failed to delete notification');
    }
  }

  /// Clear all notifications
  Future<void> clearAll() async {
    final token = await _getToken();
    if (token == null) throw Exception('No authentication token');

    final response = await _client.delete(
      Uri.parse('$_baseUrl/clear-all'),
      headers: {'Authorization': 'Bearer $token'},
    );

    if (response.statusCode != 200) {
      throw Exception('Failed to clear notifications');
    }
  }

  /// Initialize sample notifications for new users
  Future<void> initSampleNotifications() async {
    final token = await _getToken();
    if (token == null) return;

    try {
      await _client.post(
        Uri.parse('$_baseUrl/init-samples'),
        headers: {'Authorization': 'Bearer $token'},
      );
    } catch (e) {
      // Silently fail
    }
  }

  /// Create a custom notification
  Future<void> createNotification({
    required String title,
    required String message,
    String type = 'tip',
  }) async {
    final token = await _getToken();
    if (token == null) return;

    try {
      await _client.post(
        Uri.parse(_baseUrl),
        headers: {
          'Authorization': 'Bearer $token',
          'Content-Type': 'application/json',
        },
        body: jsonEncode({'title': title, 'message': message, 'type': type}),
      );
    } catch (e) {
      // Silently fail
    }
  }

  /// Notification when meditation is completed
  Future<void> notifyMeditationCompleted(
    String sessionName,
    int minutes,
  ) async {
    await createNotification(
      title: 'Meditation Complete! 🧘',
      message:
          'You completed a $minutes-minute session of "$sessionName". Great work on prioritizing your mental health!',
      type: 'achievement',
    );
  }

  /// Notification when a therapeutic game is played
  Future<void> notifyGamePlayed(String gameName) async {
    await createNotification(
      title: 'Game Completed! 🎮',
      message:
          'You finished playing "$gameName". These exercises can help reduce stress and improve mindfulness.',
      type: 'achievement',
    );
  }

  /// Notification for daily motivation
  Future<void> sendDailyMotivation() async {
    final tips = [
      'Remember to take deep breaths when feeling stressed.',
      'Small steps lead to big changes. Keep going!',
      'Your mental health matters. Take time for yourself today.',
      'Gratitude can transform your perspective. What are you thankful for?',
      'Movement is medicine. Consider a short walk today.',
    ];
    final tip = tips[DateTime.now().day % tips.length];

    await createNotification(
      title: 'Daily Motivation 💪',
      message: tip,
      type: 'tip',
    );
  }
}
